database/src/main/java/com/firebase/ui/database/paging/FirebaseDataSource.java

```diff
@@ -0,0 +1,182 @@
+
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.Query;
+import com.google.firebase.database.ValueEventListener;
+import com.google.firebase.database.annotations.NotNull;
+
+import java.util.ArrayList;
+
+/**
+ * Data source to power a {@link FirebaseRecyclerPagingAdapter}.
+ *
+ * Note: although loadInitial, loadBefore, and loadAfter are not called on the main thread by the
+ *       paging library, we treat them as if they were so that we can facilitate retry without
+ *       managing our own thread pool or requiring the user to pass us an executor.
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public class FirebaseDataSource<T> extends PageKeyedDataSource<String, T> {
+
+    private Query mQuery;
+    private Class<T> mClass;
+    private ArrayList<String> mKeyList;
+
+    private static final String TAG = "FirebaseDataSource";
+
+    private final MutableLiveData<LoadingState> mLoadingState = new MutableLiveData<>();
+    private final MutableLiveData<DatabaseError> mError = new MutableLiveData<>();
+    private final MutableLiveData<ArrayList<String>> mKeyLiveData = new MutableLiveData<>();
+
+    private static final String WRONG_DATA_PATH_STATUS = "WRONG_PATH";
+    private static final String WRONG_DATA_PATH_MESSAGE = "WRONG DATA PATH";
+    private static final String WRONG_DATA_PATH_DETAILS = "Wrong Data Path is given. Data Child Not Found !";
+
+    public static class Factory<T> extends DataSource.Factory<String, Class<T>> {
+
+        private final Query mQuery;
+        private final Class mClass;
+
+        public Factory(@NonNull Query query, @NotNull Class<T> modelClass) {
+            mQuery = query;
+            mClass = modelClass;
+        }
+
+        @Override
+        @NonNull
+        public DataSource<String, Class<T>> create() {
+            return new FirebaseDataSource(mQuery, mClass);
+        }
+    }
+
+    FirebaseDataSource(Query mQuery, Class<T> modelClass){
+        this.mQuery = mQuery;
+        this.mClass = modelClass;
+        mKeyList = new ArrayList<>();
+    }
+
+    @Override
+    public void loadInitial(@NonNull LoadInitialParams<String> params, @NonNull final LoadInitialCallback<String, T> callback) {
+        // Set initial loading state
+        mLoadingState.postValue(LoadingState.LOADING_INITIAL);
+
+        Query mInitQuery = mQuery.limitToFirst(params.requestedLoadSize);
+        mInitQuery.addListenerForSingleValueEvent(new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
+                if (dataSnapshot.exists()) {
+                    ArrayList<T> mDataList = new ArrayList<T>();
+                    for (DataSnapshot ds : dataSnapshot.getChildren()) {
+                        T data = ds.getValue(mClass);
+                        String key = ds.getKey();
+                        mKeyList.add(key);
```

Comment:
See above comment.  If we were dealing only in `DataSnapshot` here we would not have to keep separate lists of the keys and the data, which will reduce bugs and complexity.